### PR TITLE
scplan: rewrite stage builder

### PR DESCRIPTION
**Pull request by postamar**
The motivation behind this change was to make the declarative schema
changer planner more predictable. For example, for a given schema
change, we want the plan produced in the PostCommitPhase to be identical
to that produced earlier in the StatementPhase, minus the stages which
have already executed. This has the nice consequence of EXPLAIN (DDL)
being a lot more accurate and useful.

With this in mind, I rewrote the stage builder in scstage to build
stages in a constructive manner, not reliant on any topological ordering
of the nodes in the graph. Once the stages are built the planner now
thoroughtly validates them to check for inconsistencies.

Doing so required a slew of supporting changes which are hard to isolate
from this one:
- A new PostCommitNonRevertiblePhase has been spun out of the
  PostCommitPhase, to capture the non-revertible post-commit ops and
  shunt them as late in the plan as possible.
- The schema changer job is now correctly marked as non-cancellable when
  reaching the PostCommitNonRevertiblePhase.
- The schema changer job ops and the handling of the schema changer job
  ID have been changed in light of everything that precedes.
- The deprules and scopt rules have been augmented and/or corrected,
  along with some opgen tweaks.
- The scop.Ops type has been removed.

Release note: None

### CODE REVIEW

File: pkg/sql/schemachanger/scplan/plan.go [hunk 1]
```diff
@@ -72,32 +70,34 @@ func MakePlan(initial scpb.State, params Params) (p Plan, err error) {
...
-	p.stages = scstage.CollapseStages(p.stages)
-	p.stages, p.JobID = scstage.AugmentStagesForJob(p.stages, params.ExecutionPhase, params.JobIDGenerator)
-	if err = scstage.ValidateStages(p.stages); err != nil {
-		return p, errors.WithAssertionFailure(errors.Wrapf(err, "invalid improved execution plan"))
+	g, err = scopt.OptimizeGraph(g)
+	if err != nil {
+		panic(errors.Wrapf(err, "mark op edges as no-op"))
 	}
-	return p, nil
+	return g
```
Comment (ajwerner) [inline pkg/sql/schemachanger/scplan/plan.go:102]: It upsets me that we don't validate at this point. I suspect it upsets you too. The reasoning is clear, I don't think that the resultant graph would be valid. You can see this in the plan tests where we just now leave the columns in a non-terminal state when dropping the table. I'm fine deferring these changes, but at this point, this call is the scariest part of this project. The builder might have implementation issues, but they don't upset me philosophically like this part does.

File: pkg/sql/schemachanger/scplan/scstage/build.go [hunk 2]
```diff
@@ -21,249 +21,420 @@ import (
...
+	opEdges    []*scgraph.OpEdge
+}
+
+type currentTargetState struct {
+	n *scpb.Node
+	e *scgraph.OpEdge
+
+	// opFlag is true iff this stage already includes an op edge with ops for this
+	// target.
+	opFlag bool
```
Comment (ajwerner) [inline pkg/sql/schemachanger/scplan/scstage/build.go:183]: can we call this something more verbose? I find myself having to think a bit every time I read it `alreadyEmittedOpsForCurrentStage`?

**Review by ajwerner — COMMENTED** [review]
This logic is much cleaner 🙌 . I like it a lot. My feedback is a sad mix of little things and bigger things that may be out of scope for this PR.

**Review by postamar — COMMENTED** [review]
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-:-MrNjZgVCdgztIDNoLgy:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scgraphviz/graphviz.go, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfpl4-r3-86:-MrOaVEHEHKW7EzbHX0A:b-9755fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scgraphviz/graphviz.go#L86)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This change is smart and makes me happy. 
</blockquote></details>

Thank you!
___
*[pkg/sql/schemachanger/scplan/plan.go, line 80 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfplL-r3-80:-MrNSgNH6GILHcW88ZAd:b-qk5ill) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/plan.go#L80)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`errors.NewAssertionErrorWithWrappedErrf`? 
</blockquote></details>

I'm addressing this by shifting the burden on the caller and having `scgraphviz.DecorateErrorWithPlanDetails` wrap this as an assertion failure.
___
*[pkg/sql/schemachanger/scplan/plan.go, line 102 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfplL-r3-102:-MrNZjWj6NbIeNgUFt6P:b2zfwm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/plan.go#L102)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It upsets me that we don't validate at this point. I suspect it upsets you too. The reasoning is clear, I don't think that the resultant graph would be valid. You can see this in the plan tests where we just now leave the columns in a non-terminal state when dropping the table. I'm fine deferring these changes, but at this point, this call is the scariest part of this project. The builder might have implementation issues, but they don't upset me philosophically like this part does. 
</blockquote></details>

We talked about this offline and agreed that we would revisit this after the release. This, but also opgen, and basically all of how we define this graph. We've agreed that it would be nice to have motivating cases that would guide our remodelling of the graph generation.
___
*[pkg/sql/schemachanger/scplan/scstage/build.go, line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfplQ-r3-183:-MrNfcRK2svSn5LU4Ltk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/scstage/build.go#L183)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can we call this something more verbose? I find myself having to think a bit every time I read it `alreadyEmittedOpsForCurrentStage`? 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/scstage/build.go, line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfplQ-r3-186:-MrNi-r42JS8WQ3OmVD5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/scstage/build.go#L186)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The return value of this function makes it sort of confusing. In particular because `currentTargetState{n: n}` means we're not adding this mode to this stage. What if we refactored it into a bool predicate function followed by constructing the currentTargetState. I think it'd read more clearly. 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/scstage/build.go, line 353 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfplQ-r3-353:-MrNg_WhBGoDwSdjlBBr:b-a2dhuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/scstage/build.go#L353)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We can remove this todo I think 
</blockquote></details>

I'd rather leave it if you don't mind, I find it provides useful commentary.
___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 146 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfplM-r3-146:-MrNjVTCCPfMWs00ubL5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/plan_test.go#L146)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This is cute, and clever enough to warrant commentary. 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-Mr91ZCA_yVeG-jGfpkk-r3-13:-MrNmmnV8mpdxO-62IcG:bkbbjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1dad2cef93696d4a4228c7ee316fb72b0397856/pkg/sql/schemachanger/scplan/testdata/alter_table#L13)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It's interesting to think about what the dropping of the name here means. I can't decide whether or not it makes sense. I don't think that it does. My sense is that we should stick this on the edge which makes this index `Validated` and we should stick the edge which sets the index name on the new primary index to run when that index hits `Public`. We then should consider logic that takes indexes without names and gives them some temp name and then fail validation if a public index has a temp name. 
</blockquote></details>

I've reworked this to use placeholder names when creating column/index descriptors in the declarative schema changer. I've made the rules looser than what you suggested, we can always tighten them up later on when we need to. At least now, the renaming _does_ behave correctly I believe. This led me to find a bug in our CREATE INDEX implementation in scbuild.

**Review by ajwerner — APPROVED** [review]
I think you're very close to eliminating the `new_primary_key` nonsense. I won't be mad if you don't do it now. Other than that, :lgtm: 

> Reviewed 22 of 62 files at r3, 20 of 22 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-:-MrOpHlq9yZ5g8a2pUsn:b-bgtd8d)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/schemachanger/scplan/scstage/stage.go, line 31 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-MrOmPk4E7UmUpvw83nV:-MrOmPk4E7UmUpvw83nW:brgtd4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/355b3343537dc1fe5338b88716d1c712cfc24a30/pkg/sql/schemachanger/scplan/scstage/stage.go#L31)):*
> ```Go
> type Stage struct {
> 	Before, After          scpb.State
> 	EdgeOps, ExtraOps      []scop.Op
> ```

nit: this could use commentary 
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 13 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73994#-MrOo2WgCSi8H1FnObUx:-MrOo2WgCSi8H1FnObUy:b-l5weau) ([raw file](https://github.com/cockroachdb/cockroach/blob/355b3343537dc1fe5338b88716d1c712cfc24a30/pkg/sql/schemachanger/scplan/testdata/alter_table#L13)):*
> ```
>     [ColumnName:{DescID: 54, ColumnID: 2, Name: j}, ABSENT, ADD] -> PUBLIC
>     [PrimaryIndex:{DescID: 54, IndexID: 2}, ABSENT, ADD] -> DELETE_ONLY
>     [IndexName:{DescID: 54, IndexID: 2, Name: new_primary_key}, ABSENT, ADD] -> PUBLIC
> ```

does this ever somehow become `foo_pkey`? I feel like this hack can now be safely lifted, potentially with the introduction of another rule or the tightening of some of the rules you added

**Comment by postamar** [comment]
Oh, these are good points, which hadn't really crossed my mind. I'll do so, then merge.

**Comment by postamar** [comment]
It's done. It was worth it!

bors r+

**Comment by postamar** [comment]
bors cancel

CI tripped up on expected output in the new_schema_changer logic test suite.

**Pull request merged**